include/net/socket_can.h

```diff
@@ -0,0 +1,127 @@
+/*
+ * Copyright (c) 2018 Intel Corporation.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef _SOCKET_CAN_H_
+#define _SOCKET_CAN_H_
+
+#include <device.h>
+#include <net/net_ip.h>
+#include <can.h>
+#include <zephyr/types.h>
+
+/** Max length of socket packet */
+#define CAN_NET_MTU (72)
+
+/**  Protocol family of socket can connection */
+#define CAN_CONTEXT_FAMILY BIT(8)
+
+/**  Context type of socket can connection */
+#define CAN_CONTEXT_TYPE BIT(9)
+
+/**  Context proto for socket can connection */
+#define CAN_CONTEXT_PROTO BIT(10)
+
+/**
+ * @brief socket addr structure for AF_CAN protocol
+ *
+ * Socket address structure for AF_CAN protocol family
+ *
+ */
+struct sockaddr_can {
+	/** protocol family for socket can */
+	sa_family_t can_family;
+	/** net_if interface associated with socket */
+	u8_t can_ifindex;
+};
+
+/**
+ * @brief ioctl call for configuring misc params in can IP.
+ *
+ * This routine provides support for configuring misc params.
+ * *
+ * @param dev       Pointer to the device structure for the driver instance.
+ * @param cmd       type of command.
+ * @param data      data related to cmd to configure parameters.
+ *
+ * @retval 0 if success
+ */
+typedef int (*socket_can_ioctl_t)(struct device *dev, u32_t cmd, void *data);
+
+/**
+ * @brief check matched CAN filter for incoming CAN message.
+ *
+ * This routine matches id with configured filters id and return true or false.
+ * *
+ * @param dev       Pointer to the device structure for the driver instance.
+ * @param pkt       incoming net_pkt
+ *
+ * @retval non zero value if filter matched otherwise returns 0.
+ */
+typedef int (*socket_can_check_filter_matched_t)(struct device *dev,
+						 struct net_pkt *pkt);
+
+/**
+ * @brief socket can driver api structure
+ * Socket CAN driver apis for iface interface functions and ioctl call.
+ *
+ */
+struct socket_can_driver_api_t {
+	/** iface interface for socket apis to send and init function*/
+	struct net_if_api iface_api;
+	/** ioctl call for socket can driver (this will be replaced by */
+	/* setsocketopt call */
+	socket_can_ioctl_t ioctl;
+	/** check matching filter for loopback messages */
+	socket_can_check_filter_matched_t check_matched_filter;
+};
+
+/**
+ * @brief socket can context structure
+ *
+ * socket can driver specific structure.
+ *
+ */
+struct socket_can_context {
+	/** id of CAN instance */
+	u32_t id;
+	/** iface interface reference structure*/
+	struct net_if *iface;
+	/** can device structure reference */
+	struct device *can_dev;
+};
+
+__syscall int socket_can_ioctl(struct device *dev, u32_t cmd, void *data);
+
+static inline int _impl_socket_can_ioctl(struct device *dev,
+					 u32_t cmd,
+					 void *data)
+{
+	const struct socket_can_driver_api_t *api = dev->driver_api;
+
+	return api->ioctl(dev, cmd, data);
+}
+__syscall int socket_can_check_filter_matched(struct device *dev,
```

Comment:
Similar comment to the above - no socket_can_check_filter_matched() function in POSIX.

Reply:
@pfalcon:   What if we get this patch in as-is and update it as soon as you'll be ready with setsockopt code?

Reply:
@laperie that's not a good idea actually

Reply:
Why do you actually need a function like socket_can_check_filter_matched?
If you don't set a filter that matches, the CAN driver will never receive a message and therefore if you get a message from the can API it passed the filter already.

Reply:
@alexanderwachter This function implements local loopback functionality mentioned in section 3.2 of https://www.kernel.org/doc/Documentation/networking/can.txt. And in https://www.can-cia.org/fileadmin/resources/documents/proceedings/2012_kleine-budde.pdf document it is mentioned to implement it via software filtering. So this function checks if sent CAN packet Id is matched with any of configured filter id on sender node then socket application will receive this locally loopbacked packet.

Reply:
Be aware that the loopback mode bit does not only check the filter and send the message back if it matches. It also set the ACK bit while sending. Basically it acknowledges its own package.
This makes a difference when your device is alone on the bus. If you do not set the loopback mode here, the controller will try to send the message forever.

Reply:
This loopback is not via can controller. It will check the matched filter in SW layer only and if matched then connection callback will be trigger with this pkt  and socket recv call will get the packet.

Reply:
Removed ioctl calls and implemented getsockopt and setsockopt APIs for configuring can mode, bit rate and filters.